server/sonar-server/src/main/java/org/sonar/server/computation/step/NewCoverageAggregationStep.java

```diff
@@ -0,0 +1,249 @@
+    CoreMetrics.NEW_UNCOVERED_CONDITIONS_KEY,
+    CoreMetrics.NEW_IT_LINES_TO_COVER_KEY,
+    CoreMetrics.NEW_IT_UNCOVERED_LINES_KEY,
+    CoreMetrics.NEW_IT_CONDITIONS_TO_COVER_KEY,
+    CoreMetrics.NEW_IT_UNCOVERED_CONDITIONS_KEY,
+    CoreMetrics.NEW_OVERALL_LINES_TO_COVER_KEY,
+    CoreMetrics.NEW_OVERALL_UNCOVERED_LINES_KEY,
+    CoreMetrics.NEW_OVERALL_CONDITIONS_TO_COVER_KEY,
+    CoreMetrics.NEW_OVERALL_UNCOVERED_CONDITIONS_KEY
+    );
+  private static final VariationBuildersFactory VARIATION_BUILDERS_FACTORY = new VariationBuildersFactory();
+
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodsHolder periodsHolder;
+  private final MeasureRepository measureRepository;
+  private final List<Metric> metrics;
+
+  public NewCoverageAggregationStep(TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder,
+    final MetricRepository metricRepository, MeasureRepository measureRepository) {
+    this.treeRootHolder = treeRootHolder;
+    this.periodsHolder = periodsHolder;
+    this.measureRepository = measureRepository;
+    this.metrics = from(AGGREGATED_METRIC_KEYS)
+      .transform(new Function<String, Metric>() {
+        @Override
+        @Nonnull
+        public Metric apply(@Nonnull String input) {
+          return metricRepository.getByKey(input);
+        }
+      }).toList();
+  }
+
+  @Override
+  public void execute() {
+    new NewCoverageAggregationComponentVisitor()
+      .visit(treeRootHolder.getRoot());
+  }
+
+  private final class NewCoverageAggregationComponentVisitor extends PathAwareVisitor<VariationBuilders> {
+    public NewCoverageAggregationComponentVisitor() {
+      super(FILE, POST_ORDER, VARIATION_BUILDERS_FACTORY);
+    }
+
+    @Override
+    protected void visitProject(Component project, Path<VariationBuilders> path) {
+      addNewMeasures(project, path.current());
+    }
+
+    @Override
+    protected void visitModule(Component module, Path<VariationBuilders> path) {
+      addNewMeasures(module, path.current());
+      updateParentVariations(path);
+    }
+
+    @Override
+    protected void visitDirectory(Component directory, Path<VariationBuilders> path) {
+      addNewMeasures(directory, path.current());
+      updateParentVariations(path);
+    }
+
+    @Override
+    protected void visitFile(Component file, Path<VariationBuilders> path) {
+      initVariationBuilders(file, path.parent());
+    }
+  }
+
+  /**
+   * Creates a new VariationBuilders instance for all Component types except FILE.
+   */
+  private static class VariationBuildersFactory extends PathAwareVisitor.SimpleStackElementFactory<VariationBuilders> {
+    @Override
+    public VariationBuilders createForAny(Component component) {
+      return new VariationBuilders();
+    }
+
+    @Override
+    public VariationBuilders createForFile(Component file) {
+      return null;
+    }
+  }
+
+  private void initVariationBuilders(Component file, VariationBuilders builders) {
+    for (Metric metric : metrics) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS135)